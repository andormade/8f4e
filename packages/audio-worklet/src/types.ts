export interface MidiModuleAddresses {
	moduleId: string;
	noteAddress: number | undefined;
	channelAddress: number | undefined;
	noteOnOffAddress: number | undefined;
	velocityAddress: number | undefined;
}

export interface MidiCCModuleAddresses {
	moduleId: string;
	valueAddress: number | undefined;
	channelAddress: number | undefined;
	selectedCCAddress: number | undefined;
}

export interface RNBOModule {
	patcherId: string;
	paramAdresses: number[];
}
