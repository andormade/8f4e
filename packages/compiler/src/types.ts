export interface CompiledModule {
	functionBody: number[];
	initialMemory: number[];
	memoryAddresses: { address: number; id: string; default?: number }[];
	moduleId: string;
	offset: number;
}

export type MemoryBuffer = Int32Array;

export interface Connection {
	fromModuleId: string;
	fromConnectorId: string;
	toModuleId: string;
	toConnectorId: string;
}

export type RelativeAddressCalculator = (nthWord: number) => number;

export type ModuleGenerator<TConfig = unknown> = (
	moduleId: string,
	offset: RelativeAddressCalculator,
	initialConfig?: TConfig
) => CompiledModule;

export type EngineConfig = Record<string, number | string | EngineConfig[]>;

export interface Engine {
	name: string;
	config: EngineConfig;
}

export type ModuleState = Record<string, any>;

export type ModuleStateExtractor<T> = (memoryBuffer: MemoryBuffer, moduleAddress: number) => T;
export type ModuleStateInserter<T> = (moduleState: T, memoryBuffer: MemoryBuffer, moduleAddress: number) => void;

export interface Module {
	id: string;
	engine: Engine;
	state: ModuleState;
}

export type MemoryAddressLookup = Record<string, number>;
