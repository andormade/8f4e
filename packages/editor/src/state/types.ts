import { CompileOptions, CompiledModuleLookup, MemoryBuffer, DataStructure } from '@8f4e/compiler';
import { Font, SpriteLookups } from '@8f4e/sprite-generator';
import { SpriteLookup } from '@8f4e/2d-engine';

export interface CodeBlock {
	code: string[];
	isOpen: boolean;
	x: number;
	y: number;
	viewport?: Viewport;
	codeBlocks?: CodeBlock[];
}

export interface Size {
	width: number;
	height: number;
}

export interface Position {
	x: number;
	y: number;
}

export type Viewport = {
	x: number;
	y: number;
};

interface ContextMenuButton {
	action?: string;
	close?: boolean;
	payload?: Record<string, unknown>;
	disabled?: boolean;
	divider?: boolean;
	title?: string;
	isSectionTitle?: boolean;
}

interface MenuItemDivider extends ContextMenuButton {
	divider: true;
	title: never;
}

export type ExtendedInstructionSet = 'debug' | 'button' | 'switch' | 'offset' | 'plot' | 'piano';

export type ContextMenuItem = ContextMenuButton | MenuItemDivider;

export type MenuGenerator = (state: State, payload?: any) => ContextMenuItem[] | Promise<ContextMenuItem[]>;

export interface ContextMenu extends Position {
	highlightedItem: number;
	itemWidth: number;
	items: ContextMenuItem[];
	open: boolean;
	menuStack: string[];
}

export interface BuildError {
	lineNumber: number;
	code: number;
	message: string;
	moduleId: string;
}

export interface Runtime {
	runner: 'audioWorklet' | 'webWorker';
	latency: number;
	renderQuantum: number;
}

export interface Compiler {
	codeBuffer: Uint8Array;
	compilationTime: number;
	cycleTime: number;
	isCompiling: boolean;
	lastCompilationStart: number;
	memoryBuffer: MemoryBuffer;
	memoryBufferFloat: Float32Array;
	memoryRef: WebAssembly.Memory;
	timerAccuracy: number;
	compiledModules: CompiledModuleLookup;
	buildErrors: BuildError[];
	compilerOptions: CompileOptions;
	allocatedMemorySize: number;
}

export interface Midi {
	outputs: MIDIOutput[];
	inputs: MIDIInput[];
}

export interface MemoryIdentifier {
	memory: DataStructure;
	showAddress: boolean;
	showEndAddress: boolean;
	bufferPointer: number;
	showBinary: boolean;
}

export interface BufferPlotter {
	width: number;
	height: number;
	x: number;
	y: number;
	minValue: number;
	maxValue: number;
	buffer: MemoryIdentifier;
	bufferLength: MemoryIdentifier | undefined;
}
export interface Switch {
	width: number;
	height: number;
	x: number;
	y: number;
	id: string;
	offValue: number;
	onValue: number;
}

export interface Debugger {
	width: number;
	height: number;
	showAddress: boolean;
	showEndAddress: boolean;
	x: number;
	y: number;
	id: string;
	memory: DataStructure;
	bufferPointer: number;
	showBinary: boolean;
}

export interface Output {
	codeBlock: CodeBlockGraphicData;
	width: number;
	height: number;
	x: number;
	y: number;
	id: string;
	calibratedMax: number;
	calibratedMin: number;
	memory: DataStructure;
}

export interface Input {
	codeBlock: CodeBlockGraphicData;
	width: number;
	height: number;
	x: number;
	y: number;
	id: string;
	wordAlignedAddress: number;
}

export interface PianoKeyboard {
	x: number;
	y: number;
	width: number;
	height: number;
	keyWidth: number;
	pressedKeys: Set<number>;
	pressedKeysListMemory: DataStructure;
	pressedNumberOfKeysMemory: DataStructure;
	startingNumber: number;
}

export interface CodeBlockGraphicData {
	width: number;
	minGridWidth: number;
	height: number;
	code: string[];
	trimmedCode: string[];
	padLength: number;
	codeToRender: number[][];
	codeColors: Array<Array<SpriteLookup | undefined>>;
	gaps: Map<number, { size: number }>;
	cursor: { col: number; row: number; x: number; y: number };
	id: string;
	positionOffsetterXWordAddress?: number;
	positionOffsetterYWordAddress?: number;
	x: number;
	y: number;
	offsetX: number;
	offsetY: number;
	gridX: number;
	gridY: number;
	isOpen: boolean;
	extras: {
		inputs: Map<string, Input>;
		outputs: Map<string, Output>;
		debuggers: Map<string, Debugger>;
		bufferPlotters: Map<string, BufferPlotter>;
		switches: Map<string, Switch>;
		buttons: Map<string, Switch>;
		pianoKeyboards: Map<number, PianoKeyboard>;
		errorMessages: Map<
			number,
			{
				message: string[];
				x: number;
				y: number;
			}
		>;
	};
	viewport: Viewport;
	parent: CodeBlockGraphicData;
	codeBlocks: Set<CodeBlockGraphicData>;
}

export type GraphicHelper = {
	spriteLookups?: SpriteLookups;
	outputsByWordAddress: Map<number, Output>;
	globalViewport: {
		width: number;
		height: number;
		roundedWidth: number;
		roundedHeight: number;
		vGrid: number;
		hGrid: number;
		borderLineCoordinates: {
			top: { startX: number; startY: number; endX: number; endY: number };
			right: { startX: number; startY: number; endX: number; endY: number };
			bottom: { startX: number; startY: number; endX: number; endY: number };
			left: { startX: number; startY: number; endX: number; endY: number };
		};
		center: { x: number; y: number };
	};
	baseCodeBlock: CodeBlockGraphicData;
	activeViewport: CodeBlockGraphicData;
	contextMenu: ContextMenu;
	draggedCodeBlock?: CodeBlockGraphicData;
	selectedCodeBlock?: CodeBlockGraphicData;
	dialog: {
		show: boolean;
		text: string;
		title: string;
		buttons: Array<{
			title: string;
			action: string;
			payload?: any;
		}>;
	};
};

export interface BinaryAsset {
	/** The id of the module that the binary data should be loaded into */
	moduleId?: string;
	/** The id of the memory that the binary data should be loaded into */
	memoryId?: string;
	/** The binary data in base64 format */
	data: string;
	/** The file name of the binary data */
	fileName: string;
}

export interface Project {
	title: string;
	author: string;
	description: string;
	codeBlocks: CodeBlock[];
	viewport: Viewport;
	sampleRate: number;
	binaryAssets: BinaryAsset[];
	audioInputBuffers: {
		moduleId: string;
		memoryId: string;
		channel: number;
		input: number;
	}[];
	audioOutputBuffers: {
		moduleId: string;
		memoryId: string;
		channel: number;
		output: number;
	}[];
}

export interface ExampleModule {
	title: string;
	author: string;
	code: string;
	tests: unknown[];
	category: string;
}

export interface Options {
	isLocalStorageEnabled: boolean;
	isDebugMode: boolean;
	localStorageId: string;
	exampleProjects: Record<string, Project>;
	exampleModules: Record<string, ExampleModule>;
}

export interface EditorSettings {
	colorScheme: string;
	font: Font;
}

export interface State {
	compiler: Compiler;
	runtime: Runtime;
	midi: Midi;
	graphicHelper: GraphicHelper;
	project: Project;
	options: Options;
	editorSettings: EditorSettings;
	compilationTime: number;
}
