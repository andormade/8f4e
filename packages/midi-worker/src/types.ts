export interface MidiModuleAddresses {
	moduleId: string;
	noteWordAddress: number | undefined;
	channelWordAddress: number | undefined;
	noteOnOffWordAddress: number | undefined;
	velocityWordAddress: number | undefined;
}

export interface MidiCCModuleAddresses {
	moduleId: string;
	valueWordAddress: number | undefined;
	channelWordAddress: number | undefined;
	selectedCCWordAddress: number | undefined;
}

export interface RNBOModule {
	patcherId: string;
	paramAdresses: number[];
}
