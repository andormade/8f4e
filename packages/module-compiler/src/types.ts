export const enum ArgumentType {
	LITERAL = 'literal',
	IDENTIFIER = 'identifier',
}

export type Argument = { type: ArgumentType.LITERAL; value: number } | { type: ArgumentType.IDENTIFIER; value: string };

export type AST = Array<{ instruction: string; arguments: Array<Argument> }>;

export enum MemoryTypes {
	DYNAMIC_ARRAY,
	STATIC_ARRAY,
	ARRAY_SIZE,
	INPUT_POINTER,
	OUTPUT,
	PRIVATE,
	NUMBER,
	ARRAY,
}

export type MemoryMap = Array<{
	type: MemoryTypes;
	address: number;
	byteAddress: number;
	id: string;
	size: number;
	default: number | number[];
	usage: number;
}>;
