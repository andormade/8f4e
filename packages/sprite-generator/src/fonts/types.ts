export enum Glyph {
	SPACE,
	DOT,
	ARROW_TOP,
	ARROW_RIGHT,
	ARROW_BOTTOM,
	ARROW_LEFT,
	CONNECTOR_LEFT,
	CONNECTOR_RIGHT,
	SWITCH_KNOB,
}
