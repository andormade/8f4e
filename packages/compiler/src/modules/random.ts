import { localGet, localSet, i32const, i32load, i32store, ifelse } from '../wasm/instructions';
import { createFunctionBody, createLocalDeclaration } from '../wasm/sections';
import { Instruction, Type } from 'wasm-bytecode-utils';
import { ModuleGenerator } from '../types';
import { I16_SIGNED_LARGEST_NUMBER } from '../consts';

export const enum Memory {
	OUTPUT,
}

const enum Locals {
	RANDOM,
	FEEDBACK,
	__LENGTH,
}

/* Linear-feedback shift register style random generator. */
const random: ModuleGenerator = function (moduleId, offset) {
	const functionBody = createFunctionBody(
		[createLocalDeclaration(Type.I32, Locals.__LENGTH)],
		[
			...i32const(offset(Memory.OUTPUT)),
			...[
				...i32const(offset(Memory.OUTPUT)),
				...i32load(),
				...localSet(Locals.RANDOM),

				...localGet(Locals.RANDOM),
				...i32const(1 << 0),
				Instruction.I32_AND,
				...localSet(Locals.FEEDBACK),

				...localGet(Locals.RANDOM),
				...i32const(1 << 3),
				Instruction.I32_AND,
				...i32const(3),
				Instruction.I32_SHR_S,
				...localGet(Locals.FEEDBACK),
				Instruction.I32_XOR,
				...localSet(Locals.FEEDBACK),

				...localGet(Locals.RANDOM),
				...i32const(1 << 4),
				Instruction.I32_AND,
				...i32const(4),
				Instruction.I32_SHR_S,
				...localGet(Locals.FEEDBACK),
				Instruction.I32_XOR,
				...localSet(Locals.FEEDBACK),

				...localGet(Locals.RANDOM),
				...i32const(1 << 6),
				Instruction.I32_AND,
				...i32const(6),
				Instruction.I32_SHR_S,
				...localGet(Locals.FEEDBACK),
				Instruction.I32_XOR,
				...localSet(Locals.FEEDBACK),

				...localGet(Locals.RANDOM),
				...i32const(1),
				Instruction.I32_SHR_U,
				...localSet(Locals.RANDOM),

				...localGet(Locals.FEEDBACK),
				0x45, // Instruction.I32_EQZ
				...ifelse(
					Type.I32,
					[...localGet(Locals.RANDOM)],
					[...localGet(Locals.RANDOM), ...i32const(0b10000000000000000000000000000000), Instruction.I32_OR]
				),
			],

			...i32const(I16_SIGNED_LARGEST_NUMBER),
			Instruction.I32_REM_S,
			...i32store(),
		]
	);

	return {
		moduleId,
		functionBody,
		offset: offset(0),
		initialMemory: [Date.now() % I16_SIGNED_LARGEST_NUMBER],
		memoryAddresses: [{ address: offset(Memory.OUTPUT), id: 'out' }],
	};
};

export default random;
