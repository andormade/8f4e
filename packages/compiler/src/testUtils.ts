import wabt from 'wabt';

import {
	createCodeSection,
	createExportSection,
	createFunctionExport,
	createFunctionSection,
	createFunctionType,
	createImportSection,
	createMemoryImport,
	createTypeSection,
} from './wasmUtils/sectionHelpers';
import { FunctionBody } from './wasmUtils/typeHelpers';
import { CompiledModule, TestModule } from './types';
import { WORD_LENGTH } from './consts';

import { compileModules, getInitialMemory } from '.';

const HEADER = [0x00, 0x61, 0x73, 0x6d];
const VERSION = [0x01, 0x00, 0x00, 0x00];

export function createSingleFunctionWASMProgram(functionBody: FunctionBody): Uint8Array {
	return Uint8Array.from([
		...HEADER,
		...VERSION,
		...createTypeSection([createFunctionType([], [])]),
		...createImportSection([createMemoryImport('js', 'memory')]),
		...createFunctionSection([0x00]),
		...createExportSection([createFunctionExport('test', 0x00)]),
		...createCodeSection([functionBody]),
	]);
}

export function setInitialMemory(memory: Int32Array, module: CompiledModule): void {
	const initialMemory = getInitialMemory(module);
	for (let i = 0; i < initialMemory.length; i++) {
		memory[i] = initialMemory[i];
	}
}

export async function createTestModule(sourceCode: string): Promise<TestModule> {
	let allocatedMemoryForTestData = 0;

	const module: CompiledModule = compileModules([{ code: sourceCode.split('\n') }])[0];
	const program = createSingleFunctionWASMProgram(module.functionBody);

	const memoryRef = new WebAssembly.Memory({ initial: 1 });
	const memoryBuffer = new Int32Array(memoryRef.buffer);

	const { instance } = await WebAssembly.instantiate(program, {
		js: {
			memory: memoryRef,
		},
	});

	const reset = () => {
		setInitialMemory(memoryBuffer, module);

		// Clear the test data that was out of the module's scope.
		for (let i = module.memoryWordSize; i < module.memoryWordSize + allocatedMemoryForTestData; i++) {
			memoryBuffer[i] = 0;
		}
		allocatedMemoryForTestData = 0;
	};

	reset();

	const test = instance.exports.test as CallableFunction;

	const wat: string = await new Promise(resolve => {
		wabt().then(_wabt => {
			const module = _wabt.readWasm(program, {});
			resolve(module.toText({}));
		});
	});

	const memoryGet = (address: string | number): number | undefined => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);

			if (!memoryItem) {
				return;
			}

			return memoryBuffer[memoryItem.relativeWordAddress];
		}
		return memoryBuffer[address];
	};

	const memorySet = (address: string | number, value: number | number[]): void => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);
			if (!memoryItem) {
				return;
			}

			if (typeof value === 'number') {
				memoryBuffer[memoryItem.relativeWordAddress] = value;
			} else {
				for (let i = 0; i < value.length; i++) {
					memoryBuffer[memoryItem.relativeWordAddress + i] = value[i];
				}
			}
			return;
		}

		if (typeof value === 'number') {
			memoryBuffer[address] = value;
		} else {
			for (let i = 0; i < value.length; i++) {
				memoryBuffer[address] = value[i];
			}
		}
	};

	const getByteAddress = (address: string | number): number | undefined => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);
			if (!memoryItem) {
				return;
			}

			return memoryItem.byteAddress;
		}

		return address * WORD_LENGTH;
	};

	const allocMemoryForPointer = (address: string | number): number => {
		const firstFreeAddress = module.memoryWordSize + allocatedMemoryForTestData;
		memorySet(address, firstFreeAddress * WORD_LENGTH);
		allocatedMemoryForTestData++;
		return firstFreeAddress;
	};

	// @ts-ignore
	memoryBuffer.get = memoryGet;
	// @ts-ignore
	memoryBuffer.set = memorySet;
	// @ts-ignore
	memoryBuffer.allocMemoryForPointer = allocMemoryForPointer;
	// @ts-ignore
	memoryBuffer.byteAddress = getByteAddress;

	return {
		// @ts-ignore
		memory: memoryBuffer,
		test,
		reset,
		wat,
		program,
		memoryMap: module.memoryMap,
		ast: module.ast,
	};
}
