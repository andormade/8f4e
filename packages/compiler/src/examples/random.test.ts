import random from './random.asm';

import { createTestModule } from '../testUtils';
import { TestModule } from '../types';

let testModule: TestModule;

const fixtures = [-30826, -15413, 25062, 12531, 6265, -29637];

describe('random', () => {
	beforeAll(async () => {
		testModule = await createTestModule(random());
		const { memory } = testModule;
		memory.set('seed', 69420);
	});

	test('if the generated ast matches with the snapshot', () => {
		expect(testModule.ast).toMatchSnapshot();
	});

	test('if the wat code matches with the snapshot', () => {
		expect(testModule.wat).toMatchSnapshot();
	});

	test('if the generated memory map matches with the snapshot', () => {
		expect(testModule.memoryMap).toMatchSnapshot();
	});

	test.each(fixtures)('given the seed 69420, the output should be %p', output => {
		const { memory, test } = testModule;

		test();
		expect(memory.get('out')).toBe(output);
	});
});
