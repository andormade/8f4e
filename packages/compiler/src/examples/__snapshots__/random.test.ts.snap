// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`random if the generated ast matches with the snapshot 1`] = `
Array [
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "module",
    "lineNumber": 1,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "seed",
      },
      Object {
        "type": "literal",
        "value": 69420,
      },
    ],
    "instruction": "memory",
    "lineNumber": 3,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "out",
      },
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "memory",
    "lineNumber": 4,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "local",
    "lineNumber": 6,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "feedback",
      },
    ],
    "instruction": "local",
    "lineNumber": 7,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "out",
      },
    ],
    "instruction": "push",
    "lineNumber": 9,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 10,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "localGet",
    "lineNumber": 12,
  },
  Object {
    "arguments": Array [],
    "instruction": "equalToZero",
    "lineNumber": 14,
  },
  Object {
    "arguments": Array [],
    "instruction": "if",
    "lineNumber": 15,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "seed",
      },
    ],
    "instruction": "push",
    "lineNumber": 16,
  },
  Object {
    "arguments": Array [],
    "instruction": "else",
    "lineNumber": 17,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "push",
    "lineNumber": 18,
  },
  Object {
    "arguments": Array [],
    "instruction": "end",
    "lineNumber": 19,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 20,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "push",
    "lineNumber": 23,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 1,
      },
    ],
    "instruction": "push",
    "lineNumber": 24,
  },
  Object {
    "arguments": Array [],
    "instruction": "and",
    "lineNumber": 25,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "push",
    "lineNumber": 26,
  },
  Object {
    "arguments": Array [],
    "instruction": "shiftRight",
    "lineNumber": 27,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "feedback",
      },
    ],
    "instruction": "push",
    "lineNumber": 28,
  },
  Object {
    "arguments": Array [],
    "instruction": "xor",
    "lineNumber": 29,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "feedback",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 30,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "push",
    "lineNumber": 33,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 4,
      },
    ],
    "instruction": "push",
    "lineNumber": 34,
  },
  Object {
    "arguments": Array [],
    "instruction": "and",
    "lineNumber": 35,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 2,
      },
    ],
    "instruction": "push",
    "lineNumber": 36,
  },
  Object {
    "arguments": Array [],
    "instruction": "shiftRight",
    "lineNumber": 37,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "feedback",
      },
    ],
    "instruction": "push",
    "lineNumber": 38,
  },
  Object {
    "arguments": Array [],
    "instruction": "xor",
    "lineNumber": 39,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "feedback",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 40,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "push",
    "lineNumber": 43,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 8,
      },
    ],
    "instruction": "push",
    "lineNumber": 44,
  },
  Object {
    "arguments": Array [],
    "instruction": "and",
    "lineNumber": 45,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 3,
      },
    ],
    "instruction": "push",
    "lineNumber": 46,
  },
  Object {
    "arguments": Array [],
    "instruction": "shiftRight",
    "lineNumber": 47,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "feedback",
      },
    ],
    "instruction": "push",
    "lineNumber": 48,
  },
  Object {
    "arguments": Array [],
    "instruction": "xor",
    "lineNumber": 49,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "feedback",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 50,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "push",
    "lineNumber": 53,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 32,
      },
    ],
    "instruction": "push",
    "lineNumber": 54,
  },
  Object {
    "arguments": Array [],
    "instruction": "and",
    "lineNumber": 55,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 5,
      },
    ],
    "instruction": "push",
    "lineNumber": 56,
  },
  Object {
    "arguments": Array [],
    "instruction": "shiftRight",
    "lineNumber": 57,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "feedback",
      },
    ],
    "instruction": "push",
    "lineNumber": 58,
  },
  Object {
    "arguments": Array [],
    "instruction": "xor",
    "lineNumber": 59,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "feedback",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 60,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "push",
    "lineNumber": 63,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 1,
      },
    ],
    "instruction": "push",
    "lineNumber": 64,
  },
  Object {
    "arguments": Array [],
    "instruction": "shiftRightUnsigned",
    "lineNumber": 65,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 66,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "feedback",
      },
    ],
    "instruction": "push",
    "lineNumber": 68,
  },
  Object {
    "arguments": Array [],
    "instruction": "equalToZero",
    "lineNumber": 69,
  },
  Object {
    "arguments": Array [],
    "instruction": "if",
    "lineNumber": 70,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "push",
    "lineNumber": 71,
  },
  Object {
    "arguments": Array [],
    "instruction": "else",
    "lineNumber": 72,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "push",
    "lineNumber": 73,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 2147483648,
      },
    ],
    "instruction": "push",
    "lineNumber": 74,
  },
  Object {
    "arguments": Array [],
    "instruction": "or",
    "lineNumber": 75,
  },
  Object {
    "arguments": Array [],
    "instruction": "end",
    "lineNumber": 76,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 77,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&out",
      },
    ],
    "instruction": "push",
    "lineNumber": 79,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "random",
      },
    ],
    "instruction": "push",
    "lineNumber": 80,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "I16_SIGNED_LARGEST_NUMBER",
      },
    ],
    "instruction": "push",
    "lineNumber": 81,
  },
  Object {
    "arguments": Array [],
    "instruction": "remainder",
    "lineNumber": 82,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 83,
  },
]
`;

exports[`random if the generated memory map matches with the snapshot 1`] = `
Map {
  "seed" => Object {
    "byteAddress": 0,
    "default": 69420,
    "id": "seed",
    "lineNumber": 3,
    "relativeWordAddress": 0,
    "type": 1,
    "wordSize": 1,
  },
  "out" => Object {
    "byteAddress": 4,
    "default": 0,
    "id": "out",
    "lineNumber": 4,
    "relativeWordAddress": 1,
    "type": 1,
    "wordSize": 1,
  },
}
`;

exports[`random if the wat code matches with the snapshot 1`] = `
"(module
  (type (;0;) (func))
  (import \\"js\\" \\"memory\\" (memory (;0;) 1))
  (func (;0;) (type 0)
    (local i32 i32)
    i32.const 4
    i32.load
    local.set 0
    local.get 0
    i32.eqz
    if (result i32)  ;; label = @1
      i32.const 0
      i32.load
    else
      local.get 0
    end
    local.set 0
    local.get 0
    i32.const 1
    i32.and
    i32.const 0
    i32.shr_s
    local.get 1
    i32.xor
    local.set 1
    local.get 0
    i32.const 4
    i32.and
    i32.const 2
    i32.shr_s
    local.get 1
    i32.xor
    local.set 1
    local.get 0
    i32.const 8
    i32.and
    i32.const 3
    i32.shr_s
    local.get 1
    i32.xor
    local.set 1
    local.get 0
    i32.const 32
    i32.and
    i32.const 5
    i32.shr_s
    local.get 1
    i32.xor
    local.set 1
    local.get 0
    i32.const 1
    i32.shr_u
    local.set 0
    local.get 1
    i32.eqz
    if (result i32)  ;; label = @1
      local.get 0
    else
      local.get 0
      i32.const -2147483648
      i32.or
    end
    local.set 0
    i32.const 4
    local.get 0
    i32.const 32767
    i32.rem_s
    i32.store)
  (export \\"test\\" (func 0)))
"
`;
